### DISCUSSION TITLE
CoreWCF with CSLA 5.5.3 in a .Net 7 project

### DISCUSSION THREAD

**Discussion by oskarjf** [MarimerLLC/csla#3448] [Questions]
Hi

I'm upgrading a WCF project to CoreWCF and I can see there is currently a implementation of IWcfPortal. That interface doesn't seem to be available in .Net Core.  What should I use instead?

/Oskar

**Comment by rockfordlhotka** [comment]
Microsoft dropped WCF starting with .NET Core and afaik they don't plan to reimplement it in modern .NET.

Fortunately, there's the open-source CoreWCF project, so there's at least some option.

I only have time to focus on things that Microsoft ships as part of .NET. If someone were to contribute an OpenWCF data portal channel I would accept that pull request (though only for CSLA 8 in the current `main` branch - I'm not adding features to CSLA 5, 6, or 7).

Implementing a data portal channel isn't terribly difficult in CSLA 5 and later. For example, here's the CSLA 5 gRPC channel:

https://github.com/MarimerLLC/csla/tree/v5.x/Source/Csla.Channels.Grpc

Basically there's one file for the client that implements `IDataPortalProxy` and calls the server. And one file for the server that listens on the chosen transport protocol (WCF), decodes the inbound message, and calls the server-side data portal. I think the current implementation in `main` is a little simpler than it was in CSLA 5, because now the server-side data portal is injected into the listener.

**Reply by rockfordlhotka** [reply]
I should also say that a WCF channel for modern CSLA should do exactly what the gRPC and HTTP channels do - use the MobileFormatter to serialize/deserialize the object graph, and pass a binary stream (probably `byte[]`) through the WCF transport.

Back in the day, when .NET integrated/supported WCF, the WCF channel used `DataContractSerializer` (really a variant) to do the serialization. Because that's no longer supported as part of .NET, we're no longer making any effort to ensure that any serialization works _except_ for MobileFormatter. To be clear: only MobileFormatter should be assumed to work.

This too should simplify implementation of a new OpenWCF channel, because the contract between client and server is just a byte array.

**Reply by rockfordlhotka** [reply]
fwiw, there's already a backlog item for an OpenWCF channel 😄 https://github.com/MarimerLLC/csla/issues/1183
